bool ending
string filename
string folder